bson/_cbsonmodule.c

```diff
@@ -1049,6 +1115,10 @@ static int _write_element_to_buffer(PyObject* self, buffer_t buffer,
         }
         *(pymongo_buffer_get_buffer(buffer) + type_byte) = 0x09;
         return buffer_write_int64(buffer, (int64_t)millis);
+    } else if (PyObject_TypeCheck(value, (PyTypeObject *) state->DatetimeMSType)) {
+        long long millis = millis_from_datetime_ms(value);
+        *(pymongo_buffer_get_buffer(buffer) + type_byte) = 0x09;
+        return buffer_write_int64(buffer, (int64_t)millis);
```

Comment:
What should happen when millis_from_datetime_ms returns an error? Hint: let's add a test case for it.

Reply:
Added test cases are in `test_c_millis_from_datetime_ms` in `test_bson`.

Reply:
This is still not resolved. The problem is that even when `millis_from_datetime_ms` fails, we still call buffer_write_int64 with millis==-1 as if the call succeeded. I suggest refactoring `millis_from_datetime_ms` to accept an "out" parameter so that an error can be returned more easily, for example:
```c
long long millis;
if (!millis_from_datetime_ms(value, &millis)) {
    return 0;
}
*(pymongo_buffer_get_buffer(buffer) + type_byte) = 0x09;
return buffer_write_int64(buffer, (int64_t)millis);
```